velox/functions/lib/Re2Functions.cpp

```diff
@@ -1193,120 +1453,202 @@ class PatternStringIterator {
+      currentKind = SubPatternKind::kLiteralString;
+    }
 
-      if (iterator.isSingleCharWildcard()) {
-        ++singleCharacterWildcardCount;
-      } else {
-        ++anyCharacterWildcardCount;
-      }
+    // New sub pattern occurs.
+    if (currentKind != previousKind && cursor > 0) {
+      subPatternKinds.push_back(previousKind);
+      subPatternRanges.push_back(
+          {currentSubPatternStart, cursor - currentSubPatternStart});
+      currentSubPatternStart = cursor;
+    }
 
-      if (!iterator.isPreviousWildcard()) {
-        ++numWildcardSequences;
-      }
+    // Advance the cursor.
+    std::string_view currentChar = iterator.current();
+    cursor += currentChar.size();
+    previousKind = currentKind;
 
-      // Mark the end of the fixed pattern.
-      if (fixedPatternStart != -1 && fixedPatternEnd == -1) {
-        fixedPatternEnd = currentStart - 1;
-      }
-    } else {
-      // Record the first fixed pattern start.
-      if (fixedPatternStart == -1) {
-        fixedPatternStart = currentStart;
-      } else {
-        // This is not the first fixed pattern, not supported, so fallback.
-        if (iterator.isPreviousWildcard()) {
-          return PatternMetadata{PatternKind::kGeneric, 0};
-        }
-      }
-    }
+    os << iterator.current();
   }
-
-  // The pattern end may not been marked if there is no wildcard char after
-  // pattern start, so we mark it here.
-  if (fixedPatternStart != -1 && fixedPatternEnd == -1) {
-    fixedPatternEnd = patternLength - 1;
+  std::string unescapedPattern = os.str();
+
+  // Handle the last sub-pattern.
+  subPatternKinds.push_back(previousKind);
+  subPatternRanges.push_back(
+      {currentSubPatternStart, cursor - currentSubPatternStart});
+
+  const auto numSubPatterns = subPatternKinds.size();
+  SubPatternStats literalStats{SubPatternKind::kLiteralString};
+  SubPatternStats singleCharWildcardStats{SubPatternKind::kSingleCharWildcard};
+  SubPatternStats anyCharsWildcardStats{SubPatternKind::kAnyCharsWildcard};
+
+  // Collect the sub-pattern stats.
+  for (auto i = 0; i < numSubPatterns; i++) {
+    const auto& subPatternKind = subPatternKinds[i];
+    switch (subPatternKind) {
+      case SubPatternKind::kLiteralString:
+        literalStats.update(i);
+        break;
+      case SubPatternKind::kSingleCharWildcard:
+        singleCharWildcardStats.update(i);
+        break;
+      case SubPatternKind::kAnyCharsWildcard:
+        anyCharsWildcardStats.update(i);
+        break;
+    }
   }
 
-  // At this point pattern has max of one fixed pattern.
-  // Pattern contains wildcard characters only.
-  if (fixedPatternStart == -1) {
-    if (anyCharacterWildcardCount == 0) {
-      return PatternMetadata{
-          PatternKind::kExactlyN, singleCharacterWildcardCount};
+  // '%' can only occur at the beginning or end of the pattern if there are
+  // literal pattern inside.
+  if (literalStats.count > 0) {
+    if (anyCharsWildcardStats.count > 2 ||
```

Comment:
This check is a bit hard to parse. Maybe extract a helper method and split different conditions into separate if statements with comments.